pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/SegmentDeletionManager.java

```diff
@@ -213,32 +210,52 @@ public void removeAgedDeletedSegments(int retentionInDays) {
         return;
       }
 
-      AgeFileFilter fileFilter = new AgeFileFilter(DateTime.now().minusDays(retentionInDays).toDate());
-      File[] directories = deletedDir.listFiles((FileFilter) DirectoryFileFilter.DIRECTORY);
-      // Check that the directory for deleted segments is empty
-      if (directories == null) {
-        LOGGER.warn("Deleted segment directory {} does not exist or it caused an I/O error.", deletedDir);
+      URI dataDirURI = ControllerConf.getUriFromPath(_dataDir);
+      URI deletedDirURI = ControllerConf.getUriFromPath(StringUtil.join(File.separator, _dataDir, DELETED_SEGMENTS));
+      PinotFS pinotFS = PinotFSFactory.create(dataDirURI.getScheme());
+
+      // Check that the directory for deleted segments exists
+      if (!pinotFS.isDirectory(deletedDirURI)) {
+        LOGGER.warn("Deleted segment directory {} does not exist or it is not directory.", deletedDirURI.toString());
         return;
       }
 
-      for (File currentDir : directories) {
+    try {
+      String[] deletedDirFiles = pinotFS.listFiles(deletedDirURI, false);
+      if (deletedDirFiles == null) {
+        LOGGER.warn("Deleted segment directory {} does not exist.", deletedDirURI.toString());
+        return;
+      }
+
+      for (String currentDir : deletedDirFiles) {
+        URI currentURI = ControllerConf.getUriFromPath(currentDir);
         // Get files that are aged
-        Collection<File> targetFiles = FileUtils.listFiles(currentDir, fileFilter, null);
-        // Delete aged files
-        for (File f : targetFiles) {
-          if (!f.delete()) {
-            LOGGER.warn("Cannot remove file {} from deleted directory.", f.getAbsolutePath());
+        String[] targetFiles = pinotFS.listFiles(currentURI, false);
+        int numFilesDeleted = 0;
+        for (String targetFile : targetFiles) {
+          URI targetURI = ControllerConf.getUriFromPath(targetFile);
+          Date dateToDelete = DateTime.now().minusDays(retentionInDays).toDate();
+          if (pinotFS.lastModified(targetURI) < dateToDelete.getTime()) {
+            if (!pinotFS.delete(targetURI)) {
+              LOGGER.warn("Cannot remove file {} from deleted directory.", targetURI.toString());
+            } else {
+              numFilesDeleted ++;
+            }
           }
         }
-        // Delete directory if it's empty
-        if (currentDir.list() != null && currentDir.list().length == 0) {
-          if (!currentDir.delete()) {
-            LOGGER.warn("The directory {} cannot be removed. The directory may not be empty.", currentDir.getAbsolutePath());
+
+        if (numFilesDeleted == targetFiles.length) {
+          // Delete directory if it's empty
+          if (!pinotFS.delete(currentURI)) {
+            LOGGER.warn("The directory {} cannot be removed.", currentDir);
```

Comment:
What is the expected behavior of pinotFS if directory is not empty? For example, it is possible that a user deletes a segment and a file is moved into the directory after we initialize targetFiles. In that case, although we _think_ the directory should be empty, it will not be.
Can you document expected behavior in pinotFS? If it is a force remove, then we have a problem where a file is purged long before it is due to be.